hal/src/rtl872x/exflash_hal.cpp

```diff
@@ -176,6 +392,99 @@ class ExFlashLock {
     bool threading_;
 };
 
+static int writeInBestModeWithinPageBoundaries(uint32_t addr, uint8_t* buf, size_t size) {
+    uint32_t address = addr;
+    uint8_t cmd = GENERIC_FLASH_CMD_PP;
+    uint32_t mode = RTK_SPIC_DATA_WIDTH_MODE_SINGLE;
+
+    if (flash_init_para.FLASH_cur_bitmode == SpicQuadBitMode) {
+        // Quad mode enabled
+        if (flash_init_para.FLASH_Id == FLASH_ID_GD) {
```

Comment:
Where is `flash_init_para` populated? in KM0 on boot? Then copied to KM4 in `Set_System()`?
```
    _memcpy((void *)&flash_init_para, &flash_init_para_km0, sizeof(FLASH_InitTypeDef));
```
How are `FLASH_cur_bitmode` and `FLASH_Id` populated? They must be valid by the time exflash_hal is run right?

Reply:
It's all done in `rtl8721dlp_flashclk.c`, see the changeset https://github.com/particle-iot/realtek_ambd_sdk_public/compare/ede161210878fb0ee996f6edd0c4f81892109317...a6b17b249f9134ae5fe001d884558c88bf78a872#diff-798e71d08c3dcef6271de7277133461c1e34a9317546ece0b6702c722afbc0e7

and then yes, copied from KM0 to KM4. `FLASH_Id` is read from the flash (manufacturer id), bit mode is tested in `flash_rx_mode_switch()`